src/main/java/patching/PatchManager.java

```diff
@@ -148,57 +180,111 @@ public void run() {
+        }, PATCH_TIMEOUT_MILLIS);
     }
 
-    @Override
-    public void handleNotification(Notification notification) {
-        FileChangeNotification fileChangeNotif = (FileChangeNotification) notification.getData();
-        long fileID = notification.getResourceID();
-
-        if (!batchingByFile.containsKey(fileID)) {
-            BatchingControl batchingControl = new BatchingControl();
-            batchingByFile.put(fileID, batchingControl);
-            batchingControl.maxVersionSeen = -1;
-        }
-        BatchingControl batchingCtrl = batchingByFile.get(fileID);
+    // This has to be in a separate thread, so that we can have a queue, to make sure
+    // notifications are applied in the order they are received. Otherwise the waiting for
+    // no active change requests could come out in the wrong order.
+    private void runNotificationHandlerThread() {
+        new Thread(() -> {
+            Thread.currentThread().setName("PatchManagerNotificationHandler");
+
+            while (true) {
+                Notification notification = null;
+                try {
+                    notification = notificationHandlerQueue.take();
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
 
-        synchronized (batchingCtrl.patchBatchingQueue) {
-            String[] changes = fileChangeNotif.changes;
-
-            for (int i = 0; i < changes.length; i++) {
-                Patch patch = new Patch(changes[i]);
-
-                logger.debug(String.format("PatchManager-Notification: Transforming %s against doneQueue %s", patch, batchingCtrl.patchBatchingQueue).replace("\n", "\\n") + "\n");
-                logger.debug(String.format("PatchManager-Notification: Transforming %s against batchingQueue %s", patch, batchingCtrl.patchDoneQueue).replace("\n", "\\n") + "\n");
-
-                // In event that the
-                Iterator<Patch> itr = batchingCtrl.patchDoneQueue.iterator();
-                while(itr.hasNext()){
-                    Patch donePatch = itr.next();
-                    if(donePatch.getBaseVersion() >= patch.getBaseVersion()){
-                        patch = patch.transform(donePatch);
-                    } else {
-                        // remove ones we don't care about anymore.
-                        itr.remove();
+                FileChangeNotification fileChangeNotif = (FileChangeNotification) notification.getData();
+                long fileID = notification.getResourceID();
+
+                BatchingControl batchingCtrl = getBatchingControl(fileID);
+                if (batchingCtrl.activeChangeRequest) {
+                    synchronized (batchingCtrl) {
+                        while (batchingCtrl.activeChangeRequest) {
+                            try {
+                                batchingCtrl.wait();
+                            } catch (InterruptedException e) {
+                                e.printStackTrace();
```

Comment:
done. this is a wonky case, that has no "correct" solution, i don't think, but the best we can do is just apply the patch anyways.